From b4ada9f12ebcc778dff5a63cbdf594c22cbb75f5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jan=20=C4=8Cern=C3=BD?= <jcerny@redhat.com>
Date: Tue, 16 May 2023 14:48:20 +0200
Subject: [PATCH 43/43] Fix other occurences of oscap_htable_add

If the hash table owns its elements, its responsible for freeing
them. The callers of oscap_htable_add rely on the fact that the
elements will be later freed by oscap_htable_free. However, if
oscap_htable_add fails to insert the elements to the table,
then the caller needs to free them.
---
 src/DS/rds.c              | 5 ++++-
 src/OVAL/oval_generator.c | 5 ++++-
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/src/DS/rds.c b/src/DS/rds.c
index 5ec98daa2..d2553b2db 100644
--- a/src/DS/rds.c
+++ b/src/DS/rds.c
@@ -888,7 +888,10 @@ int ds_rds_create(const char* sds_file, const char* xccdf_result_file, const cha
 				result = -1;
 				oscap_source_free(oval_source);
 			} else {
-				oscap_htable_add(oval_result_sources, *oval_result_files, oval_source);
+				if (!oscap_htable_add(oval_result_sources, *oval_result_files, oval_source)) {
+					result = -1;
+					oscap_source_free(oval_source);
+				}
 			}
 			oval_result_files++;
 		}
diff --git a/src/OVAL/oval_generator.c b/src/OVAL/oval_generator.c
index 267f83037..e423a9551 100644
--- a/src/OVAL/oval_generator.c
+++ b/src/OVAL/oval_generator.c
@@ -171,7 +171,10 @@ void oval_generator_update_timestamp(struct oval_generator *generator)
 
 void oval_generator_add_platform_schema_version(struct oval_generator *generator, const char *platform, const char *schema_version)
 {
-	oscap_htable_add(generator->platform_schema_versions, platform, oscap_strdup(schema_version));
+	char *schema_version_dup = oscap_strdup(schema_version);
+	if (!oscap_htable_add(generator->platform_schema_versions, platform, schema_version_dup)) {
+		free(schema_version_dup);
+	}
 }
 
 
-- 
2.34.1

